### PR TITLE
Units python exports

### PR DESCRIPTION
**Pull request by nvaytet**
This is the first pass at exposing the units to python.
The `Unit` class was simply exposed to python, and a `units` submodule was created, containing the basic units `m`, `s`, `dimensionless` etc... Operations between different units are possible in python, and are handled by the operators of the `Unit` class.

The following Python code
```Python
import numpy as np
import dataset as ds

var1 = ds.Variable(ds.Data.Value, [ds.Dim.X], np.arange(4))
var2 = ds.Variable(ds.Coord.X, [ds.Dim.X], np.arange(4))
var1.name = "SomeNeutronData"

print("var1.name", var1.name)
print("var1.unit", var1.unit)
print("var2.unit", var2.unit)

u = ds.units.angstrom
print("u = ", u)
var1.unit = u
print("var1.unit", var1.unit)

var1.unit = ds.units.m * ds.units.m
print("var1.unit", var1.unit)

var1.unit = ds.units.counts / ds.units.us
print("var1.unit", var1.unit)

var1.unit = ds.units.counts / ds.units.m
print("var1.unit", var1.unit)
```
now produces the following output:
```
var1.name SomeNeutronData
var1.unit dimensionless
var2.unit m
u =  AA
var1.unit AA
var1.unit m^2
var1.unit counts us^-1
Traceback (most recent call last):
  File "python_units.py", line 37, in <module>
    var1.unit = ds.units.counts / ds.units.m
RuntimeError: Unsupported unit as result of division counts/m
```

### CODE REVIEW

File: exports/dataset.cpp [hunk 1]
```diff
@@ -517,6 +538,8 @@ PYBIND11_MODULE(dataset, m) {
       .def_property_readonly("tag", &Variable::tag)
       .def_property("name", [](const Variable &self) { return self.name(); },
                     &Variable::setName)
+      .def_property("unit", [](const Variable &self) { return self.unit(); },
```
Comment (SimonHeybrock) [inline exports/dataset.cpp]: Not sure why the lambda is needed here, doesn't `&Variable::unit()` work? Same for `&Variable::name()` above?

File: python/test/test_units.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import unittest
+
+from dataset import *
+import numpy as np
+
+class TestUnits(unittest.TestCase):
+    def test_create_unit(self):
+        u = units.angstrom
+        self.assertEqual(repr(u), "AA")
```
Comment (SimonHeybrock) [inline python/test/test_units.py]: Is it possible to get an actual Angstrom symbol here (using UTF8?)

File: python/test/test_units.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    def test_variable_unit(self):
+        var1 = Variable(Data.Value, [Dim.X], np.arange(4))
+        var2 = Variable(Coord.X, [Dim.X], np.arange(4))
+        self.assertEqual(repr(var1.unit), "dimensionless")
+        self.assertEqual(repr(var2.unit), "m")
```
Comment (SimonHeybrock) [inline python/test/test_units.py]: Can you also do `self.assertEqual(var2.unit, units::m)`? If not, is comparison missing from the `Unit` exports?

I'd put a tests of `repr` as a separate test and use normal comparison wherever possible.

File: python/test/test_units.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        self.assertEqual(repr(var1.unit), "AA")
+        var1.unit = units.m * units.m
+        self.assertEqual(repr(var1.unit), "m^2")
+        var1.unit = units.counts / units.us
+        self.assertEqual(repr(var1.unit), "counts us^-1")
+        with self.assertRaisesRegex(RuntimeError, "Unsupported unit as result of division counts/m"):
+            var1.unit = units.counts / units.m
+        var1.unit = ds.units.m / ds.units.m * ds.units.counts
+        self.assertEqual(repr(var1.unit), "counts")
+        with self.assertRaisesRegex(RuntimeError, "Unsupported unit as result of multiplication counts*m"):
```
Comment (SimonHeybrock) [inline python/test/test_units.py]: Maybe explain in a comment that this is just due to (currently) unsupported intermediate units, not part of the "supported" set.

File: exports/dataset.cpp (hunk 1 above)
Comment (nvaytet) [reply exports/dataset.cpp]: yes, `&Variable::unit()` works, but it doesn't work for the `name()` above. Probably something to do with the not so trivial implementation of `name()` (https://github.com/mantidproject/dataset/blob/master/src/variable.h#L186)

File: python/test/test_units.py (hunk 2 above)
Comment (nvaytet) [reply python/test/test_units.py]: yes

**Comment by nvaytet** [comment]
> Some comments below. Could you please also fix and re-enable these two tests that had temporarily been disabled due to lack of unit support:
> 
> https://github.com/mantidproject/dataset/blob/master/python/test/test_dataset.py#L270
> 
> https://github.com/mantidproject/dataset/blob/master/python/test/test_dataset.py#L394

How were you thinking of setting the units for a dataset?
`dataset[Data.Value, "data"].unit = units.counts` ?

**Pull request merged**